### PR TITLE
v0.1.8: feat/amplitude analytics

### PR DESCRIPTION
**Pull request by carlomigueldy**
- feat: define amplitude types for adding generic args support
- feat: add web3WalletAddress in IAccountContext
- feat: track on strategy changes implemented
- feat: more track functions exposed for various types of ui events
- feat: track opened market page
- feat: track share event
- refactor: amplitude track funcs
- refactor: auth context interface
- feat: track strategy, outcome, & pricing preset change/click events
- feat: track investor cabinet page opened
- feat: track deposit page opened
- feat: track wallet address copied
- feat: track login clicked event
- feat: track explore markets clicked
- feat: tarck sign out clicked
- feat: track explore markets clicked & sign out clicked event

### CODE REVIEW

File: src/services/Amplitude.tsx [hunk 1]
```diff
@@ -23,14 +44,56 @@ export const AmplitudeProvider = ({ children }: PropsWithChildren) => {
...
+      value,
+    })
+  }
+
+  function trackCopied(event: CopiedEvent, value?: CopiedEventValue) {
+    amplitudeTrack(String(event), {
+      account: accountMetadata,
+      value,
+    })
+  }
```
Comment (maxschnaider) [inline src/services/Amplitude.tsx:87]: @carlomigueldy do you believe we should start using regular functions declaration rather than arrow?

File: src/types/amplitude.ts
Comment (maxschnaider) [inline src/types/amplitude.ts:1]: @carlomigueldy beautiful stuff with types. just a thought I have in mind:
I've recently started using `types` dir to common types only and declaring service-related types together with implementation (in this case, it would be in `services/Amplitude.tsx`). IMO it's more useful and more related to DDD. curious to hear your thoughts?

File: src/types/amplitude.ts [hunk 2]
```diff
@@ -0,0 +1,101 @@
+/** eslint-ignore @typescript-eslint/no-empty-interface */
+
+import { IAccountContext } from '@/services'
```
Comment (maxschnaider) [inline src/types/amplitude.ts:3]: also, I'm curious if we need to add an import from another service, as it kinda breaks incapsulation rules, and also it's hard to follow all the imports queue over time. eg. Amplitude service imports Account service through `types` dir, but in this case we need to manually track if Account service is exported before Amplitude etc. the possible solution may be to incapsulate such common types (used across a few different services) into `types` dir itself, so `types` dir files won't import anything from services. wdyt?

File: src/components/layouts/HeaderProfileMenuMobile.tsx [hunk 3]
```diff
@@ -168,7 +170,10 @@ export const HeaderProfileMenuMobile = ({ ...props }: IButton) => {
               fontWeight={'normal'}
               colorScheme={'transparent'}
               justifyContent={'start'}
-              onClick={() => router.push('/')}
+              onClick={() => {
+                router.push('/')
+                trackOpened(OpenedEvent.PageOpened, 'Explore Markets Clicked')
```
Comment (maxschnaider) [inline src/components/layouts/HeaderProfileMenuMobile.tsx:175]: @carlomigueldy just a bit contradictory to me: `onClick` != `PageOpened`. as there's a chance that Explore Markets page won't open in this scenario (eg. some 4**/5** issues etc). should we rather track `ClickedEvent` in this case, and track `PageOpened` with `useEffect(() => {}, [])` directly in the page file?

File: src/services/Amplitude.tsx (hunk 1 above)
Comment (carlomigueldy) [reply src/services/Amplitude.tsx:87]: I did went with arrow functions than regular functions, but with arrow functions for some weird reason it didn't allow me to create generic arguments. Then at some point I decided to remove generic arguments, and now we're seeing just regular functions. 

But I still want to keep things consistent, so I will try again if it allows me to put generic arguments in arrow functions.

File: src/components/layouts/HeaderProfileMenuMobile.tsx [hunk 4]
```diff
@@ -181,7 +186,10 @@ export const HeaderProfileMenuMobile = ({ ...props }: IButton) => {
               h={'40px'}
               colorScheme={'transparent'}
               justifyContent={'start'}
-              onClick={() => signOut()}
+              onClick={() => {
+                signOut()
+                trackClicked(ClickedEvent.LogoutClicked)
```
Comment (maxschnaider) [inline src/components/layouts/HeaderProfileMenuMobile.tsx:191]: also, let's track clicked events before actual logic? eg. `signOut` may crash the func/app - in this case, we won't know if the user has clicked the button or not. tl;dr, IMO `ClickedEvent` should literally mean a click by user and not a related logic

File: src/types/amplitude.ts
Comment (carlomigueldy) [reply src/types/amplitude.ts:1]: @maxschnaider yes, originally I defined all the types in the same file under `Amplitudes.tsx` then it just crossed my thoughts about what if you wanted to have these type declarations in separate dir in `types`, but as mentioned now all of those should be common types. I will move them back into `Amplitudes.tsx` as orginally written. I am also in the same page with you about DDD, that's why it was quicker for me to get involved in the codebase you started

File: src/components/layouts/HeaderProfileMenuDesktop.tsx [hunk 5]
```diff
@@ -88,7 +90,10 @@ export const HeaderProfileMenuDesktop = ({ ...props }: IButton) => {
               h={'40px'}
               colorScheme={'transparent'}
               justifyContent={'start'}
-              onClick={() => signOut()}
+              onClick={() => {
+                signOut()
+                trackClicked(ClickedEvent.LogoutClicked)
```
Comment (maxschnaider) [inline src/components/layouts/HeaderProfileMenuDesktop.tsx:95]: same thoughts here

File: src/types/amplitude.ts (hunk 2 above)
Comment (carlomigueldy) [reply src/types/amplitude.ts:3]: @maxschnaider agreed, thanks for the catch. In this line of code, my intent here was to keep it consistent however it violates the encapsulation rules. I will instead define a type for it

File: src/components/layouts/Header.tsx [hunk 6]
```diff
@@ -55,7 +57,10 @@ export const Header = ({ ...props }: FlexProps) => {
           h={'40px'}
           display={{ sm: 'none', md: 'block' }}
           fontWeight={pathname == '/' ? 'bold' : 'normal'}
-          onClick={() => router.push('/')}
+          onClick={() => {
+            router.push('/')
+            trackOpened(OpenedEvent.PageOpened, 'Explore Markets Clicked')
```
Comment (maxschnaider) [inline src/components/layouts/Header.tsx:62]: ++

File: src/components/layouts/HeaderProfileMenuMobile.tsx (hunk 3 above)
Comment (carlomigueldy) [reply src/components/layouts/HeaderProfileMenuMobile.tsx:175]: @maxschnaider good points. I was actually in a dilemma in which approach to go for based on Dima's requirements. I think it would be much better also for me to propose to Dima in new changes in the requirements that would make more sense in both sides tech & non-tech.

File: src/components/auth/LogInButton.tsx [hunk 7]
```diff
@@ -1,11 +1,21 @@
...
   const { signIn } = useAuth()
 
   return (
-    <Button colorScheme={'brand'} fontWeight={'bold'} onClick={signIn} {...props}>
+    <Button
+      colorScheme={'brand'}
+      fontWeight={'bold'}
+      onClick={() => {
+        signIn()
+        trackOpened(OpenedEvent.LoginWindowOpened)
```
Comment (maxschnaider) [inline src/components/auth/LogInButton.tsx:15]: let's maybe use https://web3auth.io/docs/sdk/pnp/web/modal/initialize#login-modal-events to track if the modal is actually opened, rather then clicked to be opened?

File: src/app/wallet/page.tsx [hunk 8]
```diff
@@ -3,19 +3,23 @@
...
 const WalletPage = () => {
   const { signIn: signInWithW3A, isLoggedIn } = useAuth()
+  const { trackOpened } = useAmplitude()
 
   useEffect(() => {
     if (!isLoggedIn) {
       signInWithW3A()
     }
+
+    trackOpened(OpenedEvent.PageOpened, 'Deposit Page')
```
Comment (maxschnaider) [inline src/app/wallet/page.tsx:22]: 👌

File: src/components/auth/LogInButton.tsx (hunk 7 above)
Comment (carlomigueldy) [reply src/components/auth/LogInButton.tsx:15]: Much better, didn't know about those events from web3auth, thanks for sending a reference

File: src/services/Amplitude.tsx (hunk 1 above)
Comment (carlomigueldy) [reply src/services/Amplitude.tsx:87]: Still won't allow generic arguments in arrow functions, weird

image

Yet works fine with regular functions

image

Comment (carlomigueldy) [reply src/services/Amplitude.tsx:87]: Resolved, you should add a comma right after. It looks weird but works 

image

File: src/app/portfolio/page.tsx [hunk 9]
```diff
@@ -3,12 +3,19 @@
...
 import { PortfolioHistoryTable } from '@/app/portfolio/components/PortfolioHistoryTable'
 import { MainLayout } from '@/components'
+import { useAmplitude } from '@/services'
+import { OpenedEvent } from '@/types'
 import { Box, HStack, Spacer, Stack, Text } from '@chakra-ui/react'
-import { useState } from 'react'
+import { useEffect, useState } from 'react'
 
 const PortfolioPage = () => {
+  const { trackOpened } = useAmplitude()
```
[the remark this answers is not in the document]
Comment (carlomigueldy) [reply src/app/portfolio/page.tsx:12]: Hmm, trying to understand what you mean about it. But I guess something like this?

image

File: src/app/markets/[address]/page.tsx [hunk 10]
```diff
@@ -4,11 +4,17 @@ import { MainLayout } from '@/components'
...
 import { defaultChain, markets } from '@/constants'
-import { Market } from '@/types'
+import { Market, OpenedEvent } from '@/types'
 
 const MarketPage = ({ params }: { params: { address: string } }) => {
+  const { trackOpened } = useAmplitude()
+
+  useEffect(() => {
+    trackOpened(OpenedEvent.PageOpened, 'Market Page')
+  }, [])
```
Comment (maxschnaider) [inline src/app/markets/[address]/page.tsx:16]: I believe we should track market address?

File: src/app/markets/[address]/components/TradeForm.tsx [hunk 11]
```diff
@@ -170,7 +189,12 @@ export const TradeForm = ({ ...props }: StackProps) => {
                 fontSize={'22px'}
                 fontWeight={'normal'}
                 p={2}
-                onClick={() => increaseAmount(10)}
+                onClick={() => {
+                  increaseAmount(10)
+
+                  // ? Product requirements expects +1 and +5 to be tracked
+                  trackClicked(ClickedEvent.PricePresetClicked, '+1 clicked')
+                }}
```
Comment (maxschnaider) [inline src/app/markets/[address]/components/TradeForm.tsx:197]: +1/+5 isn't relevant anymore, it's the old reqs. let's track it as +10

File: src/app/markets/[address]/components/TradeForm.tsx [hunk 12]
```diff
@@ -138,7 +152,12 @@ export const TradeForm = ({ ...props }: StackProps) => {
                 fontSize={'26px'}
                 fontWeight={'normal'}
                 p={2}
-                onClick={() => decreaseAmount(10)}
+                onClick={() => {
+                  decreaseAmount(10)
+
+                  // ? Product requirements expects +1 and +5 to be tracked
+                  trackClicked(ClickedEvent.PricePresetClicked, '-1 clicked')
```
Comment (maxschnaider) [inline src/app/markets/[address]/components/TradeForm.tsx:159]: same

File: src/app/markets/[address]/components/MarketMetadata.tsx [hunk 13]
```diff
@@ -119,7 +124,10 @@ export const MarketMetadata = ({ ...props }: StackProps) => {
                   fontWeight={'normal'}
                   colorScheme={'transparent'}
                   justifyContent={'start'}
-                  onClick={() => window.open(tweetURI, '_blank')}
+                  onClick={() => {
+                    window.open(tweetURI, '_blank')
+                    trackClicked(ClickedEvent.ShareClicked, 'X/Twitter')
```
Comment (maxschnaider) [inline src/app/markets/[address]/components/MarketMetadata.tsx:129]: same, let's track click before page is opened

**Review by maxschnaider — COMMENTED** [review]
pls RSVP on comments, thanks!

**Comment by carlomigueldy** [comment]
> pls RSVP on comments overall, an amazing job @carlomigueldy thanks!

Thanks bro! 🙌

File: src/app/markets/[address]/page.tsx (hunk 10 above)
Comment (carlomigueldy) [reply src/app/markets/[address]/page.tsx:16]: Agreed, makes much more sense to understand which market was opened

File: src/services/Amplitude.tsx (hunk 1 above)
Comment (maxschnaider) [reply src/services/Amplitude.tsx:87]: `const foo = <T extends unknown>(x: T) => x;` for old TS

File: src/components/layouts/HeaderProfileMenuMobile.tsx (hunk 3 above)
Comment (maxschnaider) [reply src/components/layouts/HeaderProfileMenuMobile.tsx:175]: absolutely agree, I it will be helpful for Dima as well to understand a potential tech points here

File: src/app/portfolio/page.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (maxschnaider) [reply src/app/portfolio/page.tsx:12]: yeah that's exactly what I meant. it's not crucial ofc, just a small thing, but it helps to save time especially when code gets complicated

[the remark this answers is not in the document]
Comment (carlomigueldy) [reply src/app/portfolio/page.tsx:12]: Agreed++
